### ISSUE TITLE
F.gelu(approximate='none'): 1+erf catastrophic cancellation — up to 57% relative error in fp32 around x≈−5, zeros the negative tail

### ISSUE THREAD

**Issue by nkapreTT** [pytorch/pytorch#187806]
### 🐛 Describe the bug

`torch.nn.functional.gelu(x, approximate='none')` evaluates `0.5 * x * (1 + erf(x/√2))`. For negative `x`, `erf(x/√2) → −1`, so `1 + erf(...)` subtracts two near-equal magnitudes and loses significant digits (catastrophic cancellation). In fp32 this produces large relative errors at ordinary inputs (x ≈ −5) and collapses the output to exactly `0.0` from x ≈ −6 onward; in fp64 it zeros out from x ≈ −9.

The mathematically identical `0.5 * x * erfc(-x/√2)` computes the small value directly and stays faithful (`erfc` exists precisely to avoid this `1+erf` cancellation). I ran into this generating high-precision GELU reference values — the builtin disagrees badly with a 300-bit mpmath reference in the negative tail, while the `erfc` form matches it.

```python
import torch
for dtype in (torch.float32, torch.float64):
    x = torch.tensor([-4., -5., -5.5, -6., -8., -10.], dtype=dtype)
    s2 = 2.0 ** 0.5
    builtin = torch.nn.functional.gelu(x, approximate='none')
    erfc    = 0.5 * x * torch.special.erfc(-x / s2)   # equal math, cancellation-free
    print(dtype)
    for i, v in enumerate(x.tolist()):
        print(f"  x={v:6.1f}  builtin={builtin[i].item():.4e}  erfc={erfc[i].item():.4e}")
```

Output:

```
torch.float32
  x=  -4.0  builtin=-1.2648e-04  erfc=-1.2669e-04
  x=  -5.0  builtin=-1.1921e-06  erfc=-1.4333e-06
  x=  -5.5  builtin=-1.6391e-07  erfc=-1.0444e-07
  x=  -6.0  builtin=-0.0000e+00  erfc=-5.9195e-09
  x=  -8.0  builtin=-0.0000e+00  erfc=-4.9768e-15
  x= -10.0  builtin=-0.0000e+00  erfc=-7.6199e-23
torch.float64
  x=  -4.0  builtin=-1.2668e-04  erfc=-1.2668e-04
  x=  -5.0  builtin=-1.4333e-06  erfc=-1.4333e-06
  x=  -5.5  builtin=-1.0444e-07  erfc=-1.0444e-07
  x=  -6.0  builtin=-5.9195e-09  erfc=-5.9195e-09
  x=  -8.0  builtin=-4.8850e-15  erfc=-4.9768e-15
  x= -10.0  builtin=-0.0000e+00  erfc=-7.6199e-23
```

Against the true values (300-bit mpmath, which agrees with the `erfc` column), the fp32 builtin is **~17% off at x = −5**, **~57% off at x = −5.5**, and **exactly 0 (100% off) from x ≈ −6**. These are normal GELU inputs, not deep tail. `torch.erf` itself is accurate — the entire loss is in the `+1` composition.

**Why this is easy to miss:** you only see it if your reference is more precise than float64. A float32/float64 "golden" — numpy/scipy `1+erf`, or torch compared against itself — cancels identically, so the tail looks correct and standard `allclose`/ULP checks pass. It only shows up against an arbitrary-precision reference (mpmath/MPFR at a few hundred bits), which is why it slips through normal accuracy testing and, as far as I can tell, hasn't been reported.

**Why this is wrong rather than expected rounding:** the standard GELU identity has a cancellation-free form that recovers the correct value at the *same* precision (below). The tail values are representable and cheaply computable; the current formulation just discards them. This bites anyone using GELU output as a reference/golden or in extended-precision pipelines, and it's the root cause of large-ULP GELU errors in downstream low-precision kernels.

**Suggested fix (essentially free — not even a branch):** `Φ(x) = ½·erfc(−x/√2)` is exact for all x (the standard normal-CDF identity), so the exact GELU path is a one-token substitution:

```
0.5 * x * (1 + erf(x/√2))   →   0.5 * x * erfc(-x/√2)
```

Same op count (the `1 +` add becomes a unary negate on the argument), and `erfc` uses the same underlying machinery as `erf` — no extra transcendental, no branch, no LUT. Verified across x ∈ [−10, 10]: no regression for x ≥ 0 (max rel err 2.2e-16 fp64 / 1.5e-7 fp32), and the negative tail goes from 100% error to ~1.3e-14 (fp64) / ~9e-6 (fp32). The same `1+erf` cancellation is also present in `gelu_backward`, so a complete fix would update both forward and backward.

Happy to put up a PR if you're open to it.

### Versions

```
PyTorch version: 2.12.1+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A
OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Libc version: glibc-2.39
Python version: 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-216-generic-x86_64-with-glibc2.39
Is CUDA available: False
Is XPU available: False
Is XNNPACK available: True
```

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by praful-srinivasan-027** [comment]
Thanks for the detailed investigation. The numerical analysis and proposed erfc reformulation look interesting. I'd be interested in helping with a fix if maintainers think this is the right direction.

**Comment by anurag12-webster** [comment]
Hi, i'd like to help implement this, could a maintainer confirm whether replacing the formulation to `erfc` is a good direction? also happy to cordinate with @praful-srinivasan-027 to avoid duplicating work.

**Comment by lakshayg** [comment]
If you end up doing this, you can find a number of places where we have `1 ± erf(x)` using this command

```bash
git grep -i '1.*[+-].*\berf'
```

**Comment by leonardHONG** [comment]
Thanks for the clear write-up, and @lakshayg for the pointer. If nobody's actively on this yet, I'd be glad to take it — reformulating the `1 + erf` sites to the `erfc` form and adding fp32 tail-accuracy coverage. Happy to step aside if someone already has a patch in progress.
